### DISCUSSION TITLE
every 10th deletion takes longer time

### DISCUSSION THREAD

**Discussion by yulongkelly** [ClickHouse/ClickHouse#50595] [Q&A] [answered]
I create a table `ENGINE = MergeTree PRIMARY KEY (date, time) ORDER BY (date, time) SETTINGS disk='s3_cache'` with ~20 columns

I write a python script to delete a specific date and then insert new data.

The delete takes 1s normally but every 10th delete it takes ~15s or even longer

I used lightweight delete (delete from ...)

```
client.command(f"delete from {table_full_name} where date={date}")
client.insert_arrow(table_full_name, table_arrow)
```

**Comment by DanRoscigno** [comment, marked as answer]
Hi @yulongkelly , how are you measuring the time? Lightweight deletes are meant to look instantaneous from a select point of view (the rows will not be included in a select query), but the rows are masked and not deleted.  The actual deletion happens asynchronously in the background. https://clickhouse.com/docs/en/sql-reference/statements/delete#lightweight-delete-internals

**Reply by yulongkelly** [reply]
Hi @DanRoscigno,

Thanks for the reply. I think the delay is caused by the background merging/deletion.

**Reply by DanRoscigno** [reply]
exactly, Thanks for the question, and let me know if there is a way I can make this clearer in the docs (I work on the docs).  Have a good one!

**Reply by yulongkelly** [reply]
Hi @DanRoscigno, do u know if there is any way to make the merge happen later? I need to sync a large amount of data. The merge happen in the background make the process super slow

**Reply by DanRoscigno** [reply]
During your large sync are you adding and deleting records, or just adds?  I would not expect a bunch of merges unless you are adding one record at a time (you should batch, see https://clickhouse.com/docs/en/optimize/bulk-inserts and the other pages in best practices).
